Example for tasks interaction

antmendoza opened the pull request:
<!--- Note to EXTERNAL Contributors -->
<!-- Thanks for opening a PR! 
If it is a significant code change, please **make sure there is an open issue** for this. 
We work best with you when we have accepted the idea first before you code. -->

<!--- For ALL Contributors 👇 -->

## What was changed
<!-- Describe what has changed in this PR -->

## Why?
<!-- Tell your future self why have you made these changes -->

Some users look into Temporal as an alternative to BPM tools, like jBPM or Camunda. One common question is how to deal with HumanTask. This example shows a basic implementation to track task state


## Checklist
<!--- add/delete as needed --->

1. Closes <!-- add issue number here -->

2. How was this tested:
<!--- Please describe how you tested your changes/how we can test them -->

3. Any docs updates needed?
<!--- update README if applicable
      or point out where to update docs.temporal.io -->

tsurdilo commented:
hey, sorry still reviewing. ty for doing this sample!

antmendoza commented:
no hurry @tsurdilo

antmendoza commented on core/src/main/java/io/temporal/samples/taskinteraction/Task.java:
```diff
@@ -0,0 +1,72 @@
+public class Task {
+
+  private String token;
+  private Object data;
```
that is true, I am gonna change it

antmendoza commented on core/src/main/java/io/temporal/samples/taskinteraction/client/UpdateTask.java:
```diff
@@ -0,0 +1,55 @@
+import io.temporal.serviceclient.WorkflowServiceStubs;
+import java.util.Arrays;
+import java.util.List;
+
```
right now only the "task token" is required to complete the task 

https://github.com/temporalio/samples-java/pull/567/files/a92dae84caf9da16e0f5e9874e842d725706339d#diff-b73120e3efff34534d80195613afc83e29b8e8049cb9e796fdeb13147859addbR30

https://github.com/temporalio/samples-java/pull/567/files/a92dae84caf9da16e0f5e9874e842d725706339d#diff-89e67e1da37a2cdfafecf424ebbae1cd65645bdf244e414e802dafd2fdf5208aR47

antmendoza commented on core/src/main/java/io/temporal/samples/taskinteraction/client/ListOpenTasks.java:
```diff
@@ -0,0 +1,54 @@
+import io.temporal.serviceclient.WorkflowServiceStubs;
+import java.util.List;
+
+public class ListOpenTasks {
```
I agree, but also I think it could be useful to show the workflow execution internal state. Do you think we can expose this information in any other way or should I just remove it?

antmendoza replied:
`ListOpenTasks` removed, 

I am maintaining the method `getOpenTask` to query the internal state and complete the task automatically from [UpdateTask ](https://github.com/temporalio/samples-java/pull/567/files#diff-c7b4940f7928e4030c177f127608dce85fbfd270934dc427ed0f4dc2785bb5e0R49)

I can remove the method `getOpenTask`, and have the dev running the example passing the task id to [UpdateTask](https://github.com/temporalio/samples-java/pull/567/files#diff-c7b4940f7928e4030c177f127608dce85fbfd270934dc427ed0f4dc2785bb5e0R49) , instead of getting the taskid from the workflow execution using `getOpenTask`

antmendoza commented on core/src/main/java/io/temporal/samples/taskinteraction/Task.java:
```diff
@@ -0,0 +1,72 @@
+    return "Task{" + "token='" + token + '\'' + ", data=" + data + ", state=" + state + '}';
+  }
+
+  public enum STATE {
```
done , thank you!

antmendoza commented on core/src/main/java/io/temporal/samples/taskinteraction/TaskService.java:
```diff
@@ -0,0 +1,129 @@
+
+public class TaskService<R> {
+
+  private final Map<String, Task> tasks = Collections.synchronizedMap(new HashMap<>());
```
done , thank you!

tsurdilo commented:
hi, is there any outstanding things to be done on this pr? thanks

antmendoza commented:
Hi @tsurdilo this is on me, I have to remove the getOpenTask method

antmendoza commented:
This is ready to review @mfateev @tsurdilo

I have created a new workflow that manages the task lifecycle with which the initial workflow interacts.

joshmsmith commented:
I like  it, this is a nice sample to have.

tsurdilo commented:
Had a bit of a hard time following example code as in the whole time was thinking this would be good to have as a reusable interceptor that can be just applied to any workflow. Also would be nice to incorporate visibility into this as to 
be able to know how many workflows are running that have pending tasks and at which stage of approval process they are.
wdyt?

tsurdilo commented:
Approving this as think its good sample, just see previous comment about it being bit hard to understand / follow.

antmendoza commented:
Thanks @tsurdilo 

>  just see previous comment about it being bit hard to understand / follow.

Let me see if I can refactor it / make it simpler 

> Also would be nice to incorporate visibility into this as to
be able to know how many workflows are running that have pending tasks and at which stage of approval process they are.
wdyt?

👍

antmendoza commented:
@tsurdilo I have refactored the code and added some comments

This is the commit https://github.com/temporalio/samples-java/pull/567/commits/97e7a38830e013cf5b31d18d94cb4b4d748de67f

let me know if you find it easier to follow now
thanks

tsurdilo commented on core/src/main/java/io/temporal/samples/taskinteraction/WorkflowWithTasksImpl.java:
```diff
@@ -0,0 +1,76 @@
+
+    logger.info("Awaiting for the two tasks to complete");
+    // Block execution until both tasks complete
+    Promise.allOf(Arrays.asList(task1, task2)).get();
```
i think here you need to loop through promises again to check if any failed and if not call .get on each (to make sure they both actually compelted)

antmendoza replied:
I get your point and i can add it, but I don't think any of these promises can fail with the current implementation 

https://github.com/temporalio/samples-java/pull/567/files#diff-89e67e1da37a2cdfafecf424ebbae1cd65645bdf244e414e802dafd2fdf5208aR88

tsurdilo commented on core/src/main/java/io/temporal/samples/taskinteraction/client/CompleteTask.java:
```diff
@@ -0,0 +1,60 @@
+
+    Thread.sleep(200);
+    final List<Task> pendingTask = getPendingTask(workflowTaskManager);
+    System.out.println("Pending task " + pendingTask);
```
do we want system outs here?

antmendoza replied:
This is the client completing task , and this shows pending tasks.

tsurdilo commented on core/src/main/java/io/temporal/samples/taskinteraction/WorkflowTaskManagerImpl.java:
```diff
@@ -0,0 +1,113 @@
+
+    initTaskList(inputPendingTasks);
+
+    Workflow.await(() -> Workflow.getInfo().isContinueAsNewSuggested());
```
maybe im missing something here but if all tasks in task list complete 
before continueAsNewSuggested triggers this exec will running

i wonder if this can be changed to like "if we did not receive any new tasks for duration X" maybe instead?
reason is that cost of running execs on cloud so users might wanna complete this exec and use signalWithStart instead when new tasks come in? maybe im missing something so let me know

antmendoza replied:
I can add a condition so if there are no pending tasks the workflow returns/closes, WDYT @tsurdilo
